include/boost/histogram/axis/variable.hpp

```diff
@@ -79,7 +80,26 @@ class variable : public iterator_mixin<variable<Value, MetaData, Options, Alloca
   template <class It, class = detail::requires_iterator<It>>
   variable(It begin, It end, metadata_type meta = {}, options_type options = {},
            allocator_type alloc = {})
-      : metadata_base(std::move(meta)), vec_(std::move(alloc)) {
+      : metadata_base(std::move(meta))
+      , vec_([&] {
+        if (std::distance(begin, end) < 2)
+          BOOST_THROW_EXCEPTION(std::invalid_argument("bins > 0 required"));
+
+        vector_type vec(std::move(alloc));
+
+        vec.reserve(std::distance(begin, end));
+        vec.emplace_back(*begin++);
+        bool strictly_ascending = true;
+        for (; begin != end; ++begin) {
+          strictly_ascending &= vec.back() < *begin;
+          vec.emplace_back(*begin);
+        }
+        if (!strictly_ascending)
+          BOOST_THROW_EXCEPTION(
+              std::invalid_argument("input sequence must be strictly ascending"));
+        return vec;
+      }())
+      , eytzinger_layout_and_eytzinger_binary_search_(vec_) {
```

Comment:
The trick with the lambda is a bit too clever for my taste. I think it is more readable if we keep the code in the body of the constructor. This also does not provide the strong exception guarantee yet. Once the ctor provides the guarantee, that should also be stated in the docstring. `metadata_type` would have to become a forward reference for this to work.

Reply:
`eytzinger_layout_and_eytzinger_binary_search_`'s initialization needs an initialized `vec_`.
If we move the intitialization code of `vec_` into constructor body, then `eytzinger_layout_and_eytzinger_binary_search_` will default initializes into a dummy state and then initializes again using `eytzinger_layout_and_eytzinger_binary_search_t& assign(const Range& r)`, is that acceptable?